environment/src/main/java/com/sequenceiq/environment/environment/poller/StackPollerProvider.java

```diff
@@ -0,0 +1,83 @@
+package com.sequenceiq.environment.environment.poller;
+
+import com.dyngr.core.AttemptMaker;
+import com.dyngr.core.AttemptResult;
+import com.dyngr.core.AttemptResults;
+import com.dyngr.core.AttemptState;
+import com.sequenceiq.environment.environment.service.stack.StackService;
+import com.sequenceiq.flow.core.FlowConstants;
+import com.sequenceiq.flow.domain.FlowLog;
+import com.sequenceiq.flow.service.flowlog.FlowLogDBService;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import javax.ws.rs.BadRequestException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+@Component
+public class StackPollerProvider {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(StackPollerProvider.class);
+
+    private final StackService stackService;
+
+    private final FlowLogDBService flowLogDBService;
+
+    public StackPollerProvider(
+        StackService stackService,
+        FlowLogDBService flowLogDBService) {
+        this.stackService = stackService;
+        this.flowLogDBService = flowLogDBService;
+    }
+
+    public AttemptMaker<Void> stackUpdateConfigPoller(List<String> stackNames, Long envId, String flowId) {
+        List<String> mutableNamesList = new ArrayList<>(stackNames);
+        return () -> {
+            Optional<FlowLog> flowLog = flowLogDBService.getLastFlowLog(flowId);
+            if (flowLog.isPresent() && flowLog.get().getCurrentState().equals(FlowConstants.CANCELLED_STATE)) {
+                return AttemptResults.finishWith(null);
+            }
+            List<String> remaining = new ArrayList<>();
+            List<AttemptResult<Void>> results = collectStackUpdateConfigResults(mutableNamesList,
+                remaining, envId);
+            mutableNamesList.retainAll(remaining);
```

Comment:
No, it's not out of scope.  Its definition is outside the lambda.  So it gets maintained.  The function that is returned keeps getting run and referencing the scope with the mutableNamesList since it was defined in that scope.    Lambdas are weird.

Reply:
Thanks. You are right. I missed that.